### PR TITLE
[SPARK-6949] [SQL] [PySpark] Support Date/Timestamp in Column expression

### PR DESCRIPTION
**Pull request by davies**
This PR enable auto_convert in JavaGateway, then we could register a converter for a given types, for example, date and datetime.

There are two bugs related to auto_convert, see [1] and [2], we workaround it in this PR.

[1]  https://github.com/bartdag/py4j/issues/160
[2] https://github.com/bartdag/py4j/issues/161

cc @rxin @JoshRosen

### CODE REVIEW

File: python/pyspark/rdd.py [hunk 1]
```diff
@@ -2267,6 +2267,8 @@ def _prepare_for_python_RDD(sc, command, obj=None):
         # The broadcast will have same life cycle as created PythonRDD
         broadcast = sc.broadcast(pickled_command)
         pickled_command = ser.dumps(broadcast)
+    # There is a bug in py4j.java_gateway.JavaClass with auto_convert
```
Comment (rxin) [inline python/pyspark/rdd.py]: can you document what bug it is?

**Comment by rxin** [comment]
@JoshRosen can you take a look at this? I don't really know the py4j stuff.

File: python/pyspark/rdd.py (hunk 1 above)
Comment (davies) [inline python/pyspark/rdd.py]: Added a link here.

**Comment by rxin** [comment]
Looks like the change broke something in MLlib.

**Comment by rxin** [comment]
Thanks. I'm going to merge this in master.

**Pull request closed without merging**